homeassistant/components/scheduler/__init__.py

```diff
@@ -0,0 +1,506 @@
+        )
+
+
+@websocket_api.async_response
+@websocket_api.websocket_command(
+    {vol.Required("type"): "scheduler/schedules/clear_expired"}
+)
+async def async_websocket_handle_clear_expired(hass, connection, msg):
+    """Handle clearing expired schedules."""
+    await hass.data[DOMAIN].async_clear_expired_schedules()
+    connection.send_message(
+        websocket_api.result_message(msg[CONF_ID], hass.data[DOMAIN].schedules)
+    )
+
+
+@websocket_api.async_response
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "scheduler/schedules/create",
+        vol.Required(CONF_ENTITY_ID): cv.entity_id,
+        vol.Required(CONF_START_DATETIME): cv.datetime,
+        vol.Optional(CONF_END_DATETIME): cv.datetime,
+        vol.Optional(CONF_RRULE): str,
+    }
+)
+async def async_websocket_handle_create(hass, connection, msg):
+    """Handle creating a schedule."""
+    schedule = await hass.data[DOMAIN].async_create(
+        msg[CONF_ENTITY_ID],
+        msg[CONF_START_DATETIME],
+        end_datetime=msg.get(CONF_END_DATETIME),
+        rrule_str=msg.get(CONF_RRULE),
+    )
+    connection.send_message(websocket_api.result_message(msg[CONF_ID], schedule))
+
+
+@websocket_api.async_response
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "scheduler/schedules/delete",
+        vol.Required(CONF_SCHEDULE_ID): str,
+    }
+)
+async def async_websocket_handle_delete(hass, connection, msg):
+    """Handle deleting a schedule."""
+    try:
+        schedule = await hass.data[DOMAIN].async_delete(msg[CONF_SCHEDULE_ID])
+    except KeyError:
+        connection.send_message(
+            websocket_api.error_message(
+                msg[CONF_ID],
+                "schedule_not_found",
+                f"Cannot delete unknown schedule ID: {msg[CONF_SCHEDULE_ID]}",
+            )
+        )
+    else:
+        connection.send_message(websocket_api.result_message(msg[CONF_ID], schedule))
+
+
+@callback
+@websocket_api.websocket_command({vol.Required("type"): "scheduler/schedules/list"})
+def async_websocket_handle_list(hass, connection, msg):
+    """Handle getting all schedules."""
+    connection.send_message(
+        websocket_api.result_message(msg[CONF_ID], hass.data[DOMAIN].schedules)
+    )
+
+
+@websocket_api.async_response
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "scheduler/schedules/update",
+        vol.Required(CONF_SCHEDULE_ID): str,
+        vol.Optional(CONF_ENTITY_ID): cv.entity_id,
+        vol.Optional(CONF_START_DATETIME): cv.datetime,
+        vol.Optional(CONF_END_DATETIME): cv.datetime,
+        vol.Optional(CONF_RRULE): str,
+    }
+)
+async def async_websocket_handle_update(hass, connection, msg):
+    """Handle updating a schedule."""
+    msg_id = msg.pop("id")
+    msg.pop("type")
+    schedule_id = msg.pop("schedule_id")
+    data = msg
+
+    try:
+        schedule = await hass.data[DOMAIN].async_update(schedule_id, data)
+    except KeyError:
```

Comment:
What can raise a key error? we should check for that. This is too generic catching of exception.